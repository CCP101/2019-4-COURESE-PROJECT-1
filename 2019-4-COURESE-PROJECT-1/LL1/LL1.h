/*************************************************************************
	> File Name: LL1.h
	  > Author: Netcan
	  > Blog: http://www.netcan666.com
	  > Mail: 1469709759@qq.com
	  > Created Time: 2016-10-11 二 23:18:13 CST

                                                    __----~~~~~~~~~~~------___
                                   .  .   ~~//====......          __--~ ~~
                   -.            \_|//     |||\\  ~~~~~~::::... /~
                ___-==_       _-~o~  \/    |||  \\            _/~~-
        __---~~~.==~||\=_    -_--~/_-~|-   |\\   \\        _/~
    _-~~     .=~    |  \\-_    '-~7  /-   /  ||    \      /
  .~       .~       |   \\ -_    /  /-   /   ||      \   /
 /  ____  /         |     \\ ~-_/  /|- _/   .||       \ /
 |~~    ~~|--~~~~--_ \     ~==-/   | \~--===~~        .\
          '         ~-|      /|    |-~\~~       __--~~
                      |-~~-_/ |    |   ~\_   _-~            /\
                           /  \     \__   \/~                \__
                       _--~ _/ | .-~~____--~-/                  ~~==.
                      ((->/~   '.|||' -_|    ~~-/ ,              . _||
                                 -_     ~\      ~~---l__i__i__i--~~_/
                                 _-~-__   ~)  \--______________--~~
                               //.-~~~-~_--~- |-------~~~~~~~~
                                      //.-~~~--\
						神兽保佑，代码无BUG!
*************************************************************************/

#ifndef LL1_H
#define LL1_H

#include <iostream>
#include <ctype.h>
#include <vector>
#include <string>
#include <algorithm>
#include <map>
#include <set>
using namespace std;

class Prod { // 产生式
	friend class LL1;
	private:
		string prod; // 产生式
		char noTerminal; // 产生式左部非终结符名字
		set<string> selection; // 候选式集合
		set<char> Vn; // 非终结符
		set<char> Vt; // 终结符
		string cut(int i, int j) {
			return string(prod.begin() + i, prod.begin() + j);
		}
		friend bool operator == (const Prod &a, const char &c) {
			return a.noTerminal == c;
		}

		bool isValid; // 产生式是否合法

	public:
		Prod(const string &in);
		bool split(); // 分割终结符、非终结符、产生式集合、左部非终结符名字，返回分割是否成功
};

class LL1 {
	private: vector<Prod> G; // 文法G
		set<char> VN; // 非终结符
		set<char> VT; // 终结符
		map<char, set<char> > FIRST; // first集
		map<char, set<char> > FOLLOW; // follow集
		map<pair<char, char>, string> M; // 分析表
		set<char> first(const string &s);
		void follow();
		vector<char> parse; // 分析栈
		vector<char> indata; // 输入表达式栈
		void parseTable();
	public:
		bool addProd(const Prod & prod); // 添加产生式
		void info(); // 输出相关结果
		void tableInfo(); // 输出表
		void build(); // 建立first、follow集、分析表
		void showIndataStack(); // 输出输入串内容
		void showParseStack(); // 输出分析栈内容
		void loadIndata(const string &s); // 输入串入栈
		void parser(); // LL1预测分析
		void error(int step); // 错误处理
		void run(); // 运行LL1
};

#endif
