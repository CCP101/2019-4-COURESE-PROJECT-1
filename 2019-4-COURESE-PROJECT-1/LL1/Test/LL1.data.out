{
"VT": [ "(", ")", "*", "+", "-", "/", "@", "i"], 
"VN": [ "E", "F", "G", "S", "T"], 
"FIRST": [
 { "noTerminal": "E", "Terminal": [ "(", "i"] }
,{ "noTerminal": "G", "Terminal": [ "+", "-", "@"] }
,{ "noTerminal": "T", "Terminal": [ "(", "i"] }
,{ "noTerminal": "S", "Terminal": [ "*", "/", "@"] }
,{ "noTerminal": "F", "Terminal": [ "(", "i"] }], 
"FOLLOW": [
 { "noTerminal": "E", "Terminal": [ "#", ")"] }
,{ "noTerminal": "G", "Terminal": [ "#", ")"] }
,{ "noTerminal": "T", "Terminal": [ "#", ")", "+", "-"] }
,{ "noTerminal": "S", "Terminal": [ "#", ")", "+", "-"] }
,{ "noTerminal": "F", "Terminal": [ "#", ")", "*", "+", "-", "/"] }], 
"Table": {
"Header": ["(", ")", "*", "+", "-", "/", "i", "#"], 
"Body": [ {"noTerminal": "E", "production": ["TG", "", "", "", "", "", "TG", ""]}
, {"noTerminal": "G", "production": ["", "@", "", "+TG", "-TG", "", "", "@"]}
, {"noTerminal": "T", "production": ["FS", "", "", "", "", "", "FS", ""]}
, {"noTerminal": "S", "production": ["", "@", "*FS", "@", "@", "/FS", "", "@"]}
, {"noTerminal": "F", "production": ["(E)", "", "", "", "", "", "i", ""]}
] }, 
"Parser": [ { "step": 0, "parseStack": "#E", "indataStack": "i+i*i#", "production": ""}
, { "step": 1, "parseStack": "#GT", "indataStack": "i+i*i#", "production": "E->TG"}
, { "step": 2, "parseStack": "#GSF", "indataStack": "i+i*i#", "production": "T->FS"}
, { "step": 3, "parseStack": "#GSi", "indataStack": "i+i*i#", "production": "F->i"}
, { "step": 4, "parseStack": "#GS", "indataStack": "+i*i#", "production": "match i"}
, { "step": 5, "parseStack": "#G", "indataStack": "+i*i#", "production": "S->@"}
, { "step": 6, "parseStack": "#GT+", "indataStack": "+i*i#", "production": "G->+TG"}
, { "step": 7, "parseStack": "#GT", "indataStack": "i*i#", "production": "match +"}
, { "step": 8, "parseStack": "#GSF", "indataStack": "i*i#", "production": "T->FS"}
, { "step": 9, "parseStack": "#GSi", "indataStack": "i*i#", "production": "F->i"}
, { "step": 10, "parseStack": "#GS", "indataStack": "*i#", "production": "match i"}
, { "step": 11, "parseStack": "#GSF*", "indataStack": "*i#", "production": "S->*FS"}
, { "step": 12, "parseStack": "#GSF", "indataStack": "i#", "production": "match *"}
, { "step": 13, "parseStack": "#GSi", "indataStack": "i#", "production": "F->i"}
, { "step": 14, "parseStack": "#GS", "indataStack": "#", "production": "match i"}
, { "step": 15, "parseStack": "#G", "indataStack": "#", "production": "S->@"}
, { "step": 16, "parseStack": "#", "indataStack": "#", "production": "G->@"}
]
}
